### PR TITLE
Use `import std;` guarded by macro

### LINKED ISSUE #1815: C++20 module improvements
There are a few things I'd like to improve about `vulkan.cppm`, stemming from my own experience, and bug reports around.

- Using any struct that declares/exports some sort of comparison or `operator<=>` requires that the consumer `#include <compare>`, which is very unintuitive; I'd class this as a bug. Some good resources:
  - [DevComm thread: Template exports requiring importing of dependent packages (C++20 modules)](https://developercommunity.visualstudio.com/t/Template-exports-requiring-importing-of-/1425979). The suggestion made was to `export import <compare>;` but header units are generally considered a bad idea.
  - [Reddit comment from a MSVC compiler developer with further detail, making a similar suggestion](https://www.reddit.com/r/cpp/comments/1b0zem7/what_is_the_state_of_modules_in_2024/ksc8ix8/), but this time suggests using `import std;` in the _producing_ module without exporting any additional symbol. Which brings me to...
- Using `import std;` breaks badly with `import vulkan-hpp;` because the latter still relies on `#include <some_std_header>` for all its dependencies. 
  - The general recommendation is to not mix standard headers and the standard library module.
  - I propose using `import std;` guarded behind the appropriate feature macro, [`__cpp_lib_modules`](https://en.cppreference.com/w/cpp/feature_test#cpp_lib_modules). 
  - It's worth noting that the std library module has been back-ported to C++20. 
    - MSVC STL: https://github.com/microsoft/STL/issues/3945
    - LLVM libc++: https://reviews.llvm.org/D158358
    - GNU libstdc++: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106852 (in-progress)
    
I'd like to spur some discussion on this, and if it's worthwhile, start a relevant PR.

### PR DESCRIPTION
**Pull request by sharadhr**
Resolves #1815. Work in progress.

### CODE REVIEW

**Comment by qbojj** [comment]
There is one problem with this approach:
as stated in http://eel.is/c++draft/cpp#import-3:

>  If a [pp-import](http://eel.is/c++draft/cpp#nt:pp-import) is produced by source file inclusion (including by the rewrite produced when a #include directive names an importable header) while processing the [group](http://eel.is/c++draft/cpp#nt:group) of a [module-file](http://eel.is/c++draft/cpp#nt:module-file), the program is ill-formed.

so `import`s in #included files are ill-formed.

**Comment by sharadhr** [comment]
Ah... I was trying to put everything in the headers; looks like I'll have to take a leaf from @stripe2933's approach and put the `imports` in the `.cppm` itself, guarded by the macro. I'll rework this in the evening.

**Comment by sharadhr** [comment]
There's a bit of a problem here and I'd like some thoughts.

I'm defining the macro `VULKAN_HPP_CPP_STD_MODULE` in `vulkan_hpp_macros.hpp` based on the platform and compiler characteristics. This means we need `#include <vulkan/vulkan_hpp_macros.hpp>` in the global module fragment, i.e. between `module;` and `export vulkan_hpp;`. 

Then, in several of the headers (especially `vulkan.hpp`), I selectively disable `#include <stdheader>` based on the presence of `VULKAN_HPP_CPP_STD_MODULE`. 

Now, the `import std;` needs to appear *before* the various `#include` statements; otherwise the compiler complains that `std is not available`.

However, `import` statements can't appear in the global module fragment, and we have a Catch-22 situation. It can be resolved with `import <vulkanheader>` but header units are neither widely supported nor recommended.

**Comment by stripe2933** [comment]
I also worked to port the [magic_enum's module](https://github.com/Neargye/magic_enum/blob/dae6bbf16c363e9ead4e628a47fdb02956a634f3/module/magic_enum.cppm), and I think this file shows the solution:

You can include headers below to the `export module module_name;` by enclosing them with `extern "C++"`. Therefore, include `std` first, and include the Vulkan headers will solve this problem

**Comment by sharadhr** [comment]
I'm not sure that is a good resolution: [this blog post](https://vector-of-bool.github.io/2019/10/07/modules-3.html) says:

> C++ Modules is almost entirely agnostic to the existence of the preprocessor. What you’ve just done in the above code is cause every single symbol in SDL.h to become attached to my_game (with module linkage). Obviously, your module does not own the contents of SDL.h.

However, the standard *does* allow for `import` statements in the global module fragment:

[module.global.frag](https://eel.is/c++draft/module.global.frag#1)
> [Note [1](https://eel.is/c++draft/module.global.frag#note-1): Prior to phase 4 of translation, only preprocessing directives can appear in the [declaration-seq](https://eel.is/c++draft/dcl.pre#nt:declaration-seq) ([[cpp.pre]](https://eel.is/c++draft/cpp.pre))[.](https://eel.is/c++draft/module.global.frag#1.sentence-1) — end note]

[cpp.pre](https://eel.is/c++draft/cpp.pre)
> A [preprocessing directive](https://eel.is/c++draft/cpp.pre#def:preprocessing_directive) consists of a sequence of preprocessing tokens that satisfies the following constraints: At the start of translation phase 4, the first token in the sequence, referred to as a [directive-introducing token](https://eel.is/c++draft/cpp.pre#def:token,directive-introducing), begins with the first character in the source file (optionally after whitespace containing no new-line characters) or follows whitespace containing at least one new-line character, and is
[(1.1)](https://eel.is/c++draft/cpp.pre#1.1)
 a `#` preprocessing token, or
[(1.2)](https://eel.is/c++draft/cpp.pre#1.2)
 an import preprocessing token immediately followed on the same logical source line by a [header-name](https://eel.is/c++draft/lex.header#nt:header-name), `<`, [identifier](https://eel.is/c++draft/lex.name#nt:identifier), [string-literal](https://eel.is/c++draft/lex.string#nt:string-literal), or `:` preprocessing token, or
[(1.3)](https://eel.is/c++draft/cpp.pre#1.3)
 a `module` preprocessing token immediately followed on the same logical source line by an [identifier](https://eel.is/c++draft/lex.name#nt:identifier), `:`, or `;` preprocessing token, or
[(1.4)](https://eel.is/c++draft/cpp.pre#1.4)
 an `export` preprocessing token immediately followed on the same logical source line by one of the two preceding forms[.](https://eel.is/c++draft/cpp.pre#1.sentence-1)

So I believe it is safer to put the `import` in the global module fragment.

**Comment by qbojj** [comment]
There is also an option to define something like `VULKAN_HPP_EXPORT` to `export` only when compiling the module and using it with every exported declaration. With this approach, the module would look something like:
```c++
module;
#define VULKAN_HPP_EXPORT export

#include <vulkan/vulkan.h>
#include <vulkan/vulkan_hpp_macros.hpp>

#if !VULKAN_HPP_USE_STD_MODULE
// includes
#endif

export module vulkan_hpp;

#if VULKAN_HPP_USE_STD_MODULE
import std;
import std.compat;
#endif

#include <vulkan/vulkan.hpp>
// other vulkan includes
```

I've tried to go with this approach, but I had problems with `vk::DispatchLoaderDynamic` as it is defined in `vulkan_hpp_macros.hpp`, but from my [initial tries](https://github.com/qbojj/Vulkan-Hpp/tree/std-module) It should work.

**Comment by sharadhr** [comment]
I think your initial claim is misconstrued—you've quoted the standard:

> http://eel.is/c++draft/cpp#import-3:

> If a [pp-import](http://eel.is/c++draft/cpp#nt:pp-import) is produced by source file inclusion (including by the rewrite produced when a #include directive names an importable header) while processing the [group](http://eel.is/c++draft/cpp#nt:group) of a [module-file](http://eel.is/c++draft/cpp#nt:module-file), the program is ill-formed.

But `std` is not a `pp-import` (i.e. a 'header unit'); it is a named module. Headers themselves are allowed to `import` named modules. I think I'll revert to my original implementation, which is cleaner.

**Comment by qbojj** [comment]
Now that I look at it, You are right. I'm sorry for the confusion

**Comment by sharadhr** [comment]
Actually, I misconstrued that myself too. That line refers to a module 'group', which is the part between `export module vulkan_hpp;` and the end of the file, or `module :private;`. So either way, both `import <header>;` and `import namedmodule;` statements can appear in header files.

**Comment by sharadhr** [comment]
I'm not entirely sure why the CI tasks are failing so badly... @asuessenbach, any thoughts?

**Comment by asuessenbach** [comment]
> I'm not entirely sure why the CI tasks are failing so badly... @asuessenbach, any thoughts?

Just one thing: the CI on MacOS needed some update (#1942). But I don't see what might have happened on the other CIs.

**Comment by sharadhr** [comment]
Now that you mention it, I do see that build failure on MSVC, but I'm not sure they're relevant to this PR. In particular, https://github.com/KhronosGroup/Vulkan-Hpp/blob/d86b49ca041b9be51afb105668312925719845d0/tests/UniqueHandleDefaultArguments/UniqueHandleDefaultArguments.cpp#L29 needs to be:
```suggestion
- auto         uniqueSurface = vk::UniqueSurfaceKHR( surface, vk::Instance() ); 
+ auto         uniqueSurface = vk::UniqueSurfaceKHR( vk::SurfaceKHR( surface ), vk::Instance() ); 
```
because it seems the `vk::SurfaceKHR( VkSurfaceKHR )` constructor seems to have been marked `explicit` (which seems like a bug, this should only happen on 32-bit platforms).

**Comment by asuessenbach** [comment]
> because it seems the vk::SurfaceKHR( VkSurfaceKHR ) constructor seems to have been marked explicit (which seems like a bug, this should only happen on 32-bit platforms).

And that's the case with the main branch, but seems to be not the case with your branch.
Determining, if it should be explicit depends on some define from vulkan_core.h (`VK_USE_64_BIT_PTR_DEFINES`). Do you miss to include that, somehow?

**Comment by sharadhr** [comment]
Right, I've resolved it. If the macros in `vulkan_hpp_macros.hpp` depend on `vulkan.h` then it only makes sense to `#include` the latter in the former, which I've done. Now the test case passes.

**Comment by sharadhr** [comment]
Ready to merge.

**Comment by asuessenbach** [comment]
Looks great now, thanks a lot again for your contribution!

Could you please also generate the vulkansc files (by running the VulkanHppGenerator with `-api=vulkansc` as argument)?

**Comment by sharadhr** [comment]
@stripe2933 while the CI completes, if you could test this out that'd be great. Refer to https://github.com/KhronosGroup/Vulkan-Hpp/blob/41db2194537c8863927735e2500e326e65e5a0a8/tests/CppStdModule/CMakeLists.txt if you'd like to see how to set it up.

**Comment by asuessenbach** [comment]
> Is there any workaround for this?

I don't think so.
You could try to store the changes of this PR as a patch and apply that patch to the 283-build (https://github.com/KhronosGroup/Vulkan-Hpp/releases/tag/v1.3.283). But I don't know, if that actually works.

**Comment by stripe2933** [comment]
I tested this in Clang 18 by manually updating the relevant portions of the header and module files, and everything worked seamlessly. Great work, @sharadhr! If you have time, I have a couple of questions:

1. I noticed that I didn't moved `<vulkan/vulkan.h>` from `vulkan/vulkan.hpp` to `vulkan/vulkan_hpp_macros.hpp` file, yet it compiled successfully. Could you clarify why this file is moved? I think this change might not only add noticeable build overhead for every TU to include `<vulkan/vulkan_hpp_macros.hpp>` file, but also violate the meaning of the header name (most users will expect this header only contains macros about Vulkan-Hpp).
2. Could you explain the rationale behind defining `VULKAN_HPP_STD_MODULE` and `VULKAN_HPP_STD_COMPAT_MODULE` as `std` and `std.compat`, respectively? Since these are reserved identifiers in the C++ specification, I'm curious about the reasoning for their usage.

**Comment by sharadhr** [comment]
@stripe2933 thanks for testing! To answer your questions:

1. I moved `#include <vulkan/vulkan.h>` to `vulkan_hpp_macros.hpp` because the latter has the following few lines:
   
   ```c++
   // 32-bit vulkan is not typesafe for non-dispatchable handles, so don't allow copy constructors on this platform by default.
   // To enable this feature on 32-bit platforms please #define VULKAN_HPP_TYPESAFE_CONVERSION 1
   // To disable this feature on 64-bit platforms please #define VULKAN_HPP_TYPESAFE_CONVERSION 0
   #if ( VK_USE_64_BIT_PTR_DEFINES == 1 )
   #  if !defined( VULKAN_HPP_TYPESAFE_CONVERSION )
   #    define VULKAN_HPP_TYPESAFE_CONVERSION 1
   #  endif
   #endif
   ```
   and in turn, `VK_USE_64_BIT_PTR_DEFINES == 1` is defined in `vulkan_core.h` (which is included in `vulkan.h`). This means `vulkan_hpp_macros.hpp` depends on `vulkan_core.h`. 
   
   I could refactor so that the macros file only includes the core header, but there is no escaping having at least this macro in it (otherwise the macros file would be incomplete). 
   
2. I followed the style of the rest of the project where keywords and headers are re-defined with `VULKAN_HPP_*` (see what has been done with `constexpr` and `std::expected`, for instance). It is always possible that these *might* change, so it is easier to set up these macros once and re-define them to something else if necessary. Really a matter of style, I feel.

**Comment by stripe2933** [comment]
Thank you for the detailed explanation. Regarding the first question, if this issue only affects 32-bit operating systems, how about remain the `#include <vulkan/vulkan.h>` line in the `vulkan.hpp` file for 64-bit environments? Since most operating systems are 64-bit, this could help reduce the overhead.

**Comment by sharadhr** [comment]
It's the other way round—this is required for 64 B OSs (which are the majority now).

This was why the previous CI job failed.

The entire reasoning is as follows: in `vulkan.hpp` we have std header includes that need to be conditionally included (or the std module imported instead), based on a macro. That macro is in `vulkan_hpp_macros.hpp`. That in turn depends on `vulkan_core.h`.

Additionally `vulkan_hpp_macros.hpp` may be included on its own as well, and every macro in it must be well-defined or at most dependent on a user-configured value.

If you can see a way around this, though, I'll be happy to refactor it!

**Comment by stripe2933** [comment]
> What I _could_ do is copy the test and definition for `VK_USE_64_BIT_PTR_DEFINES` into `vulkan_hpp_macros.hpp`, if @asuessenbach is happy with that.

I think this is the best way to around it. [As Vulkan specification says](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_USE_64_BIT_PTR_DEFINES.html), the `vulkan_core.h` header allows the `VK_USE_64_BIT_PTR_DEFINES` definition to be overridden by the application. Re-defining the macro outside the `vulkan_core.h` file seems not dangerous anyway.

**Comment by sharadhr** [comment]
Reworked again.

**Comment by stripe2933** [comment]
- It seems `#include <version>` have to be added before the line `#if defined(__cpp_lib_modules)` in `vulkan/vulkan.cppm`.
- In `vulkan/vulkan_hpp_macros.hpp` file, I think `VULKAN_HPP_STD_MODULE` macro definition doesn't have to be enclosed with `defined(__cpp_lib_modules)`, because:
  - `VULKAN_HPP_ENABLE_STD_MODULE` does the same role.
  - It will always require user to include `<version>` header before the macro header, which looks quite troublesome.

**Comment by stripe2933** [comment]
> Not really, will it? `<version>` is already present in the macros header.

I missed that by mistake. You’re right.

---

I don’t see any other problems. Looks good to me 👍

**Review by asuessenbach — COMMENTED** [review]
Looks great, overall!
Just one issue I'd like to have resolved.

File: snippets/macros.hpp [hunk 1]
```diff
@@ -77,6 +77,20 @@
...
+#  define VULKAN_HPP_STD_COMPAT_MODULE std.compat
+#endif
+
+#if !defined( VK_USE_64_BIT_PTR_DEFINES )
+#  if defined( __LP64__ ) || defined( _WIN64 ) || ( defined( __x86_64__ ) && !defined( __ILP32__ ) ) || defined( _M_X64 ) || defined( __ia64 ) || \
+    defined( _M_IA64 ) || defined( __aarch64__ ) || defined( __powerpc64__ ) || ( defined( __riscv ) && __riscv_xlen == 64 )
+#    define VK_USE_64_BIT_PTR_DEFINES 1
+#  else
+#    define VK_USE_64_BIT_PTR_DEFINES 0
+#  endif
```
[the remark this answers is not in the document]
Comment (sharadhr) [reply snippets/macros.hpp]: There's a small side-effect: `Use64BitPtrDefines` is now a `constexpr` constant in `vulkan.hpp`.

**Comment by sharadhr** [comment]
Reworked again!

**Comment by stripe2933** [comment]
After applying this patch, I got weird build error when I construct `std::string_view` from `vk::ArrayWrapper1D<char, 256>` (e.g. from `vk::PhysicalDevice::enumerateDeviceExtensionProperties`).

> /usr/local/include/vulkan/vulkan.hpp:153:46: error: call to function 'strnlen' that is neither visible in the template definition nor found by argument-dependent lookup
>  153 |       return std::string_view( this->data(), strnlen( this->data(), N ) );

I'm using macOS 15.0 and Homebrew Clang 18.1.6. Could you check if this build error is reproduced?

**Comment by stripe2933** [comment]
I just realized that `strnlen` isn't a part of the C standard. This function really looks like it should be standard (^^;;). The function is not exported in libc++ module. I think I should replace it with `strnlen_s` or `std::find`.

File: VulkanHppGenerator.cpp [hunk 2]
```diff
@@ -16362,6 +16386,12 @@ namespace
     std::erase( rawComment, '/' );
     auto const strippedComment = trim( stripPostfix( stripPrefix( rawComment, " DEPRECATED:" ), "#define " ) );
 
+    // special case for VK_DEFINE_NON_DISPATCHABLE_HANDLE which is weird
+    if ( completeMacro.size() == 1 && completeMacro.front().find( "VK_DEFINE_NON_DISPATCHABLE_HANDLE" ) == std::string::npos )
+    {
```
[the remark this answers is not in the document]
Comment (sharadhr) [reply VulkanHppGenerator.cpp:16391]: Good point. Reworked.

**Comment by sharadhr** [comment]
> I just realized that `strnlen` isn't a part of the C standard. This function really looks like it should be standard (^^;;). The function is not exported in libc++ module. I think I should replace it with `strnlen_s` or `std::find`.

Just reproduced with MSVC 19.40 (VS 2022 17.11). `strnlen_s` doesn't work either; in fact none of the `<cstring>` or `<string.h>` functions appear to be exported. For now I have used `#include <string.h>` to mitigate this.

**Comment by sharadhr** [comment]
@asuessenbach I have rebased this on top of `main` and resolved merge conflicts. I think I'm happy with where this is; let me know if there are any further issues.

File: VulkanHppGenerator.cpp [hunk 3]
```diff
@@ -763,16 +775,20 @@ void VulkanHppGenerator::generateToStringHppFile() const
 // ignore warnings on using deprecated enum values in this header
 #if defined( __clang__ ) || defined( __GNUC__ )
 #  pragma GCC diagnostic push
-#  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #elif defined( _MSC_VER )
+#  pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```
[the remark this answers is not in the document]
Comment (sharadhr) [reply VulkanHppGenerator.cpp:779]: Hmm, that's a dodgy rebase. Let me try again...

**Comment by sharadhr** [comment]
I think there is nothing further to be done; ready to merge.

**Comment by asuessenbach** [comment]
@sharadhr Thanks a lot for your ongoing and great support here! Makes Vulkan-Hpp somewhat more future-proved.
